core/src/main/java/org/apache/druid/java/util/common/concurrent/ScheduledExecutors.java

```diff
@@ -167,6 +169,50 @@ public void run()
     );
   }
 
+  public static void scheduleAtFixedRate(CronScheduler exec, Duration rate, Callable<Signal> callable)
+  {
+    scheduleAtFixedRate(exec, rate, rate, callable);
+  }
+
+  /**
+   * Run callable once every period, after the given initial delay. Uses
+   * {@link CronScheduler} for task scheduling. Exceptions are caught and logged
+   * as errors.
+   */
+  public static void scheduleAtFixedRate(
+      final CronScheduler exec,
+      final Duration initialDelay,
+      final Duration rate,
+      final Callable<Signal> callable
+  )
+  {
+    log.debug("Scheduling periodically: %s with period %s", callable, rate);
+    Instant delayInstance = Instant.now().plusMillis(initialDelay.getMillis());
+    exec.scheduleAt(delayInstance,
```

Comment:
I think this logic: scheduling one-shot task which reschedules itself, is problematic. I think we should remove these methods (together with the `Signal` enum) because they have negative utility.

In particular, for `CronScheduler`, constant rescheduling is probably prone to some drift. Instead, `CronScheduler`'s methods like `scheduleAtFixedRate()` should be used directly from `MonitorScheduler`. The periodic task can be cancelled using the returned `Future`.

Reply:
That makes sense. I'll change this.

Reply:
Hey @leventov, do you think the below implementation for `startMonitor(final Monitor monitor)` method in `MonitorScheduler` class will suffice? Also, can this cause any inconsistency  since `scheduledFuture` is volatile and is shared amongst all the monitors?

    synchronized (lock) {
      monitor.start();
      Long rate = config.getEmitterPeriod().getMillis();
      scheduledFuture = scheduler.scheduleAtFixedRate(
          rate,
          rate,
          TimeUnit.MILLISECONDS,
          new CronTask()
          {
            @Override
            public void run(long scheduledRunTimeMillis)
            {
              try {
                if (monitor.monitor(emitter) && hasMonitor(monitor)) {
                  log.trace("Running %s (period %s)", this, rate);
                } else {
                  log.debug("Stopping rescheduling %s (delay %s)", this, rate);
                  removeMonitor(monitor);
                  while (scheduledFuture == null) {
                    Thread.sleep(1);
                  }
                  scheduledFuture.cancel(false);
                }
              } catch (Throwable e) {
                log.error(e, "Uncaught exception.");
              }
            }
          });
    }

Reply:
There should be a separate future for every monitor. I also think there should be a separate executorService for running `monitor.monitor(emitter)`, and cancelling the future from the first monitor, for two reasons:
 1. Monitor code is not guaranteed to be non-blocking and "cheap", which is the requirement of CronScheduler;
 2. You can avoid having a race condition of cancelling a future on the first iteration when it's not yet created.

Reply:
Keeping your above points in mind, I think this implementation will do.

    synchronized (lock) {
      monitor.start();
      Long rate = config.getEmitterPeriod().getMillis();
      Future<?> scheduledFuture = scheduler.scheduleAtFixedRate(
          rate,
          rate,
          TimeUnit.MILLISECONDS,
          new CronTask()
          {
            private volatile Future<Boolean> monitorFuture = null;
            @Override
            public void run(long scheduledRunTimeMillis)
            {
              try {
                if (monitorFuture != null && monitorFuture.isDone()
                    && !(monitorFuture.get() && hasMonitor(monitor))) {
                  removeMonitor(monitor);
                  monitor.getScheduledFuture().cancel(false);
                  log.debug("Stopped rescheduling %s (delay %s)", this, rate);
                  return;
                }
                
                log.trace("Running %s (period %s)", this, rate);
                monitorFuture = executor.submit(new Callable<Boolean>()
                {
                  public Boolean call()
                  {
                    try {
                      return monitor.monitor(emitter);
                    } catch (Throwable e) {
                      log.error(e, "Uncaught exception.");
                      return false;
                    }
                  }
                });
              } 
              catch (Throwable e) {
                log.error(e, "Uncaught exception.");
              }
            }
          });
      monitor.setScheduledFuture(scheduledFuture);
    }


In this:
1.  Each monitor has a separate future.
2.  Cron task is cheap, it checks a boolean condition. If condition is true it cancels the scheduling process for the 
     particular monitor, else submits a callable for monitoring to executor service 
3.  No race condition for cancelling scheduledFuture on first iteration.

Please let me know your thoughts on this.